tests/sentry/event_manager/grouping/test_assign_to_group.py

```diff
@@ -0,0 +1,451 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from collections.abc import Callable
+from contextlib import contextmanager
+from time import time
+from typing import Any, ParamSpec, TypeVar
+from unittest import mock
+
+import pytest
+
+from sentry.event_manager import _create_group
+from sentry.eventstore.models import Event
+from sentry.grouping.ingest import (
+    _calculate_primary_hash,
+    _calculate_secondary_hash,
+    find_existing_grouphash,
+)
+from sentry.models.grouphash import GroupHash
+from sentry.models.project import Project
+from sentry.testutils.helpers.eventprocessing import save_new_event
+from sentry.testutils.pytest.fixtures import django_db_all
+from sentry.testutils.skips import requires_snuba
+
+pytestmark = [requires_snuba]
+
+
+LEGACY_CONFIG = "legacy:2019-03-12"
+NEWSTYLE_CONFIG = "newstyle:2023-01-11"
+
+# TODO: Once we're on python 3.12, we can get rid of these and change the first line of the
+# signature of `make_wrapped_function_with_captured_return_values` to
+#   def make_wrapped_function_with_captured_return_values[T, **P](
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+def make_wrapped_function_with_captured_return_values(
```

Comment:
> If this doesn't seem too kludgy, I'll eventually move it into our test utils. (If I do, I might try to figure out how it can be made into/used as a legit decorator, since it seems awfully decorator-y.)

We're on the same page! You think it seems legit? If so, I'll just add it there instead. 

(The more I look at it, the more I wish the name weren't so clunky, though. I guess I could go with `wrap_function_for_capturing_return_values` (not a ton better)... Maybe just `capture_return_values`? I suppose given the extensive docstring that's probably clear enough.)

As for spying, if you want to do it without any fancy return value tracking, I've found you can just use `wraps` on the normal function. Seems to work in the same way as `jest.spyOn`, at least for the simplest cases.

Reply:
UPDATE: I did move it, and added tests, at which point it felt like it deserved its own PR: https://github.com/getsentry/sentry/pull/64856. I'll rebase this once that once merges.